File: src/android/ansight-core/src/main/kotlin/ai/ansight/runtime/HostSessionJpegCapturePolicy.kt
package ai.ansight.runtime

import org.json.JSONObject

internal data class HostSessionJpegCapturePolicy(
    val useHostCapture: Boolean,
    val source: String?,
) {
    companion object {
        const val ControlVersionPropertyName = "sessionJpegCaptureControlVersion"
        const val ControlVersion = 1
        val App = HostSessionJpegCapturePolicy(false, null)

        fun fromPayload(payload: JSONObject?): HostSessionJpegCapturePolicy {
            val capture = payload?.optJSONObject("sessionJpegCapture") ?: return App
            if (!capture.optString("mode").equals("host", ignoreCase = true)) {
                return App
            }

            return HostSessionJpegCapturePolicy(
                useHostCapture = true,
                source = capture.optionalString("source"),
            )
        }
    }
}

File: src/dotnet/Ansight.Core/Screenshot/HostSessionJpegCapturePolicy.cs
using System.Text.Json.Nodes;

namespace Ansight.Screenshot;

internal sealed record HostSessionJpegCapturePolicy(bool UseHostCapture, string? Source)
{
    public const string ControlVersionPropertyName = "sessionJpegCaptureControlVersion";
    public const int ControlVersion = 1;

    public static HostSessionJpegCapturePolicy App { get; } = new(false, null);

    public static void AddClientCapabilities(
        JsonObject payload,
        SessionJpegCaptureOptions? captureOptions)
    {
        ArgumentNullException.ThrowIfNull(payload);

        payload[ControlVersionPropertyName] = ControlVersion;
        if (captureOptions is null)
        {
            return;
        }

        payload["sessionJpegCapture"] = new JsonObject
        {
            ["maxWidth"] = captureOptions.MaxWidth
        };
    }

    public static HostSessionJpegCapturePolicy FromPayload(JsonObject? payload)
    {
        var capture = payload?["sessionJpegCapture"] as JsonObject;
        var mode = capture?["mode"]?.GetValue<string>();
        return string.Equals(mode, "host", StringComparison.OrdinalIgnoreCase)
            ? new HostSessionJpegCapturePolicy(
                true,
                capture?["source"]?.GetValue<string>())
            : App;
    }
}
